filament/src/materials/dof.mat

```diff
@@ -49,17 +44,271 @@ fragment {
+    if (isTrivialTile(tiles)) {
+        postProcess.color = vec4(0.0);
+        outAlpha = 0.0;
+        return;
+    }
+
+    vec4 foreground = vec4(0.0);
+    vec4 background = vec4(0.0);
+    float fgOpacity = 0.0;
+    float bgOpacity = 0.0;
+
+    // we use the full resolution pixel size because that's the unit the CoC is in
+    highp vec2 fullResPixelSize = 0.5 / vec2(textureSize(materialParams_foreground, 0));
+
+    // When using mipmaping the noise seems to hurt more than it helps
+    //float randomAngle = random(gl_FragCoord.xy) * (2.0 * PI);
+    //float random01 = random(gl_FragCoord.xy * vec2(7.0, 11.0));
+    //vec2  randomUniformDisk = 0.5 * sqrt(random01) * vec2(cos(randomAngle), sin(randomAngle));
+    //vec2  noise = randomUniformDisk * rcp(RING_COUNT + 0.5);
+    vec2 noise = vec2(0.0);
+
+    if (isFastTile(tiles)) {
+        // for a foreground tile, the kernel size is the largest CoC radius
+        float coc = abs(tiles.r);
+        float kernelSize = coc;
+        float mip = getMipLevel(kernelSize);
+        vec2 uvCenter = uv + noise * kernelSize * fullResPixelSize;
+
+        for (float i = 0.0; i < RING_COUNT; i += 1.0) {
+            float radius = (kernelSize / RING_COUNT) * i;
+            float count = max(1.0, RING_DENSITY * i);
+            float inc = 2.0 * PI / count;
+            vec2 t = vec2(cos(inc), sin(inc));
+            vec4 r = vec4(t.x, -t.y, t.y, t.x);
+            float firstSamplePosition = mod(i, 2.0) * 0.5 * inc;
+            vec2 p = radius * vec2(cos(firstSamplePosition), sin(firstSamplePosition));
+            for (float j = 0.0; j < count; j += 1.0) {
+                vec2 pos = uvCenter + p * fullResPixelSize;
+                vec4 s = textureLod(materialParams_foreground, pos, mip);
+                float o = cocToAlpha(coc);
+                foreground += s;
+                fgOpacity += o;
+                p = vec2(dot(p, r.xy), dot(p, r.zw));
+            }
+        }
+        foreground *= rcp(SAMPLE_COUNT);
+        fgOpacity  *= rcp(SAMPLE_COUNT);
+        foreground *= fgOpacity;
+        //foreground.b += 0.4;
+#if POST_PROCESS_OPAQUE
+        // kill the work performed above
+        foreground.a = 0.0;
+#endif
+        postProcess.color  = foreground;
+        outAlpha           = fgOpacity;
+        return;
+    }
+
+    /*
+     * Then process the slow foreground / background tiles
+     */
+
+    if (isForegroundTile(tiles)) {
+        // for a foreground tile, the kernel size is the largest CoC radius
+        float kernelSize = -tiles.g;
+        float mip = getMipLevel(kernelSize);
+        vec2 uvCenter = uv + noise * kernelSize * fullResPixelSize;
+        float c = 0.0;
+
+        // the center sample is handled separatly, since it's by itself
+        accumulateForegroundFirst(foreground, fgOpacity, c, uvCenter, 0.0, mip);
+
+        for (float i = 1.0; i < RING_COUNT; i += 1.0) {
+            float radius = (kernelSize / RING_COUNT) * i;
```

Comment:
Might be worth checking that the optimizer converts / RING_COUNT into a multiply by a constant

Reply:
yes, I checked that, at least spirvtools does it.